redis: use system jemalloc

yu-re-ka opened the pull request:
###### Description of changes

Fixes the 4K vs 16K pagesize of the build host being detected at build time in the vendored jemalloc.
This means the redis-server binary built on hydra (where the builders have 4K pagesize) running on a 16K pagesize system such as the Apple Silicon Mac.

```
[yuka@m1:~]$ nix run nixpkgs#redis
<jemalloc>: Unsupported system page size
Could not connect to Redis at 127.0.0.1:6379: Connection refused
<jemalloc>: Unsupported system page size
zmalloc: Out of memory trying to allocate 24448 bytes
Aborted (core dumped)

[yuka@m1:~]$ 
```

This also fixes cross-compilation without disabling jemalloc.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

yuyuyureka committed:
redis: use system jemalloc

yu-re-ka commented:
@ofborg build redis

yu-re-ka commented:
@ofborg build pkgsCross.armv7l-hf-multiplatform.redis

marsam commented on pkgs/servers/nosql/redis/default.nix:
```diff
@@ -20,19 +20,21 @@ stdenv.mkDerivation rec {
       sha256 = "sha256-0GMiygbO7LbL1rnuOByOJYE2BKUSI+yy6YH781E2zBw=";
     })
+    # use system jemalloc
+    (fetchpatch {
```
nitpick, I think we can use `fetchurl`. See: https://github.com/NixOS/nixpkgs/issues/240141

marsam reviewed (approved):
Would you mind adding an entry to the Release Notes?
IIRC Redis uses a custom jemalloc for active defragmentation, which is experimental and disabled by default; but some people might rely on it.

vikanezrimaya reviewed (approved):
Tested by launching both native `aarch64-linux` and cross-compiled `aarch64-linux` (using my trusty Raspberry Pi). While I don't think I have any ARM hardware with 16KB pages, I can at least verify it launches and doesn't immediately crash on...more standard and boring machines :sparkles:

yu-re-ka commented:
>  IIRC Redis uses a custom jemalloc for active defragmentation, which is experimental and disabled by default; but some people might rely on it.

Added an option to switch back to vendored jemalloc, and added a release notes entry describing this.

marsam commented on nixos/doc/manual/release-notes/rl-2111.section.md:
```diff
@@ -441,6 +441,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - `pkgs.haskell-language-server` will now by default be linked dynamically to improve TemplateHaskell compatibility. To mitigate the increased closure size it will now by default only support our current default ghc (at the moment 9.0.2). Add other ghc versions via e.g. `pkgs.haskell-language-server.override { supportedGhcVersions = [ "90" "92" ]; }`.
 
+- `pkgs.redis` is now built using the system jemalloc. This disables the experimental active defragmentation feature of redis. Users who require this features can switch back to redis' vendored version of jemalloc by setting `services.redis.package = pkgs.redis.override { useSystemJemalloc = false; };`.
```
very minorly
```suggestion
- `pkgs.redis` is now built using the system jemalloc. This disables the experimental active defragmentation feature of redis. Users who require this feature can switch back to redis' vendored version of jemalloc by setting `services.redis.package = pkgs.redis.override { useSystemJemalloc = false; };`.
```

yuyuyureka committed:
Update nixos/doc/manual/release-notes/rl-2111.section.md

Co-authored-by: Mario Rodas <marsam@users.noreply.github.com>
